src/yki/spec.clj

```diff
@@ -29,6 +28,33 @@
         (str yy "-" mm "-" dd)
         string->date)))
 
+(defn valid-email?
+  "Attempts to validate an email address *without* allowing quoting,
+   thus simplifying the validation rules quite a bit.
+   Best effort, doesn't aim to cover all cases - might be simultaneously
+   too restrictive as well as too lax.
+
+   A proper validator such as the EmailValidator class from Apache Commons
+   could also be used. However, proper validators typically perform too strict
+   domain validation, rendering test addresses such as foo@test.invalid unusable.
+   This custom implementation was chosen as a suitable middle ground.
+
+   See https://datatracker.ietf.org/doc/html/rfc3696#section-3 for reference on
+   the syntax available for email addresses."
+  [email]
+  (when-let [[local-part domain-part] (str/split email #"@")]
+    (and
+      ; Local part consists of max 64 whitelisted characters
+      (re-matches #"^[A-Za-z0-9\!#$%&'+\-\/=\?\^_`\.\{|\}~]{1,64}$" local-part)
+      ; Consecutive periods not allowed
+      (not (re-find #"\.\." local-part))
+      ; Local part must not start with a period
+      (not (str/starts-with? local-part "."))
+      ; Local part must not end with a period
+      (not (str/ends-with? local-part "."))
+      ; Domain part consists of max 255 whitelisted characters
+      (re-matches #"^[A-Za-z0-9\-\.]{1,255}" domain-part))))
```

Comment:
Jos joku syöttää vaikka `bää@pöö.com` niin feilaakohan tämä? Pitäis kai muuttaa IDNA muotoon domain jossain aiemmin jos haluais tämän toimivan oikein.

Reply:
Juu, toi tosiaan failaa. Vaikea sanoa kuinka iso ongelma tää tällä hetkellä on; tuotantokannan kaiveleminen voisi tuottaa jotain oivalluksia.

Reply:
Varmaan aika teoreettinen juttu. En usko, että käytännössä kellään on noita unicode-kirjaimia domainissa tai lokaalissa osassa.

Reply:
En tiedä osaako esim. ryhmasahkopostipalvelu-service tehdä noita IDNA-muunnoksia domainille. Etenkin jos osaa, niin voisi olla mahdollista muuttaa regexissä `A-Za-z` -> `\p{L}`.

Reply:
Tai sitten tietty voidaan muuntaa IDNA-muotoon (frontendissa tai bäkkärissä). Tai olla tekemättä mitään muutosta tähän.

Reply:
Se IDNA-muutos kaiketi tarvitsee tehdä vasta siellä lähetyspäässä eli onko se nyt sitten AWS SES vai mikä, joka sen mailin lopulta lähettää. Meidän ei ehkä kannata yrittää enkoodata noita osoitteita itse. 🤔 Lokaaliin osaanki on kai mahdollista laittaa unicodea ja se sitten enkoodataan tietyllä tavalla. Tai en ole kyllä ihan satavarma, mutta luulisi, että se lähetyspalvelu hoitaa enkoodauksen kuntoon. Jos itse enkoodataan IDNA-muotoon niin sitten menee kaikki aika vaikeaksi, kun se pitäisi sitten aina dekoodata ym.

Tuosta regexistä en osaa sanoa. Täältä kai löytyy kaikki domainissa sallitut merkit: https://unicode.org/faq/idn.html#7a Tuohon emailin validointiin on valmiita regexejäkin esim täällä: https://emailregex.com - mutta noista taas en sitten tiedä, että salliiko ne HTML:n siellä osoitteessa.

Reply:
Löysin vielä tän, oiskohan tää hyvä kompromissi sitten kuitenkin: https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address

Tuokaan ei salli unicode-emaileja, mutta jos html5 on tollainen validointi niin kai se meillekin kelpaa. 🤔

Reply:
Se virallinen speksi ja sitä myöten nuo regexit on vähän turhan sallivia omaan makuun.
Esim. `"<script>alert('moi');</script>"@test.foo` on speksin mukaan ok.

Mun ajatus regexin osalta oli että jos rajoitetaan sallittuja merkkejä (ei sallita esim. kommentteja tai quotattavaa sisältöä) lokaaliosaan, niin varmaan voidaan löytää sopiva tasapaino ymmärrettävyyden ja kattavuuden osalta.

Reply:
Njoo. Ehkä nuo unicode-emailit voi vielä unohtaa ja mennään tällä. Muuttaa sitten jos tulee asiakaspalautetta.

Reply:
Laitoin vielä tuollaisen version jossa sallitaan Unicodesta normaaleja kirjaimia (\p{L}, kts. https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html#ucc) mutta ei esim. emojeja. Tein vielä testikeissin joissa on muutamia hyväksyttyjä ja hylättyjä tapauksia dokumentoitu.